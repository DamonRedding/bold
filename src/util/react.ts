import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
