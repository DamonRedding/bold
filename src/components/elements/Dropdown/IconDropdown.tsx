import * as React from 'react'
import { Overlay } from 'react-overlays'

import { withStyles, WithStylesProps } from '../../../styles'
import { DropdownItem, DropdownMenu } from '../Dropdown/DropdownMenu'
import { Icons } from '../Icon/generated/Icons'
import { Icon } from '../Icon/Icon'
import { OverlayContent } from '../Overlay/OverlayContent'

export interface OptionItem {
    title: string
    onClick(): any
}

export interface IconDropdownProps extends WithStylesProps {
    icon?: Icons
    options: OptionItem[]
}

export interface IconDropdownState {
    show: boolean
}

@withStyles
export class IconDropdown extends React.Component<IconDropdownProps, IconDropdownState> {

    static defaultProps: Partial<IconDropdownProps> = {
        icon: 'dots',
    }

    private trigger

    constructor(props: IconDropdownProps) {
        super(props)
        this.state = {
            show: false,
        }
    }

    render() {
        const { icon, options, css } = this.props
        const styles = {
            container: {
                position: 'relative',
                marginLeft: '0.5rem',
            },
            button: {
                fontWeight: 'bold',
            },
            icon: {
                marginLeft: '0.25rem',
            },
        }

        return (
            <span className={css(styles.container)}>
                <a className={css(styles.button)} onClick={this.toggleShow} ref={this.triggerRef}>
                    <span className={css(styles.icon)}><Icon icon={icon} /></span>
                </a>
                <Overlay
                    show={this.state.show}
                    placement='bottom'
                    container={this}
                    target={this.trigger}
                    rootClose={true}
                    onHide={this.toggleShow}
                >
                    <OverlayContent>
                        <DropdownMenu>
                            {options.map((option, idx) => (
                                <DropdownItem key={idx} onClick={this.handleClick(option)}>{option.title}</DropdownItem>
                            ))}
                        </DropdownMenu>
                    </OverlayContent>
                </Overlay>
            </span>
        )
    }

    handleClick = (option: OptionItem) => (e) => {
        option.onClick()
        this.toggleShow()
    }

    toggleShow = () => {
        this.setState({ show: !this.state.show })
    }

    triggerRef = (elem) => {
        this.trigger = elem
    }
}
