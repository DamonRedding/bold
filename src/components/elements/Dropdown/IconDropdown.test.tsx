import { mount, render, shallow } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../test'

import { IconDropdown } from './IconDropdown'

const options = [
    {
        title: 'Option 1',
        onClick: () => alert('Option 1 selected'),
    },
    {
        title: 'Option 2',
        onClick: () => alert('Option 2 selected'),
    },
]

describe('IconDropdown', () => {
    it('deve renderizar corretamente', () => {
        const wrapper = render(withTheme(<IconDropdown options={options} icon={'triangleDown'} />))
        expect(wrapper).toMatchSnapshot()
    })
})
