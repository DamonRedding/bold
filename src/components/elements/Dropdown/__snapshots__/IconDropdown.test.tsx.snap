// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconDropdown should render correctly 1`] = `
.emotion-4 {
  position: relative;
  margin-left: 0.5rem;
}

.emotion-3 {
  font-weight: bold;
}

.emotion-2 {
  margin-left: 0.25rem;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<span
  class="emotion-4"
>
  <a
    class="emotion-3"
  >
    <span
      class="emotion-2"
    >
      <span
        class="emotion-1"
      >
        <svg
          aria-hidden="true"
          class="emotion-0"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12 17.21l6-7.22H6l6 7.22z"
            id="triangle-down"
          />
        </svg>
      </span>
    </span>
  </a>
</span>
`;
